### ISSUE TITLE
False positive "misaligned pointer dereference" on x86 Windows

### ISSUE THREAD

**Issue by glandium** [rust-lang/rust#112480]
This is a regression from #98112. I suppose it's not possible to disable this specific check only while preserving debug assertions...

The core problem is that the x86 ABI on Windows doesn't guarantee the stack alignment above 4. See for example https://developercommunity.visualstudio.com/t/vs2017-64-bit-int-alignment-problem/294259

And while some types have an alignment reported of 8 (e.g. UINT64), in practice, the C compiler will happily not align them on the stack.

So for example, this C code, compiled by MSVC for 32-bits:
```
#include <cinttypes>
extern void hoge(uint64_t*);
void foo() {
    uint64_t a;
    hoge(&a);
}
```
will produce this assembly:
```
_a$ = -8                                                ; size = 8
void foo(void) PROC                                        ; foo
        push    ebp
        mov     ebp, esp
        sub     esp, 8
        lea     eax, DWORD PTR _a$[ebp]
        push    eax
        call    void hoge(unsigned __int64 *)                        ; hoge
        add     esp, 4
        mov     esp, ebp
        pop     ebp
        ret     0
void foo(void) ENDP
```
([on godbolt](https://godbolt.org/z/5eqKrn11j))

If the stack pointer is not 8-bytes aligned when entering the function, the pointer passed to `hoge` is not going to be 8-bytes aligned.

As mentioned in the linked community post above, adding `alignas(8)` to the type definition makes the compiler align the stack:
```
#include <cinttypes>
extern void hoge(uint64_t*);
void foo() {
    alignas(8) uint64_t a;
    hoge(&a);
}
```
becomes
```
_a$ = -8                                                ; size = 8
void foo(void) PROC                                        ; foo
        push    ebx
        mov     ebx, esp
        sub     esp, 8
        and     esp, -8                             ; fffffff8H
        add     esp, 4
        push    ebp
        mov     ebp, DWORD PTR [ebx+4]
        mov     DWORD PTR [esp+4], ebp
        mov     ebp, esp
        sub     esp, 8
        lea     eax, DWORD PTR _a$[ebp]
        push    eax
        call    void hoge(unsigned __int64 *)                        ; hoge
        add     esp, 4
        mov     esp, ebp
        pop     ebp
        mov     esp, ebx
        pop     ebx
        ret     0
void foo(void) ENDP
```
([on godbolt](https://godbolt.org/z/8zP7Pezen))

Now, what this means is that if that `hoge` function is a rust FFI function, and it uses that pointer, the "misaligned pointer dereference" check is hit and panic ensues.

Real life case, for the curious:
https://github.com/servo/dwrote-rs/blob/master/src/font_file_loader_impl.rs#L116-L123

That function is called from dwrite.dll (which comes with Windows).

**Comment by glandium** [comment]
> And while some types have an alignment reported of 8 (e.g. UINT64)

Let me substantiate this:
```
#include <cinttypes>
int foo() {
    return alignof(uint64_t);
}
```
compiles to:
```
int foo(void) PROC                                        ; foo
        push    ebp
        mov     ebp, esp
        mov     eax, 8
        pop     ebp
        ret     0
int foo(void) ENDP
```
([on godbolt](https://godbolt.org/z/x1jeYhMxh))

> If the stack pointer is not 8-bytes aligned when entering the function, the pointer passed to hoge is not going to be 8-bytes aligned.

Actually, if the stack pointer *is* 8-bytes aligned when entering the function, the pointer passed to hoge is going to be unaligned (because there's a push first).

**Comment by saethlin** [comment]
> I suppose it's not possible to disable this specific check only while preserving debug assertions...

If you can pass unstable flags, currently, you can say `-Zmir-enable-passes=-CheckAlignment`.

**Comment by saethlin** [comment]
I think your report here is a duplicate of https://developercommunity.visualstudio.com/t/visual-c-alignof-is-not-conformant-on-x86/1258506

That is, this appears to be a known MSVC bug.

**Comment by glandium** [comment]
The MSVC bug would be the value returned by alignof. That doesn't change anything for the alignment check in rust.

**Comment by saethlin** [comment]
`std::mem::align_of::<u64>()` returns `8` on i686-pc-windows-msvc. Are you saying that is a bug?

**Comment by ChrisDenton** [comment]
It's complicated. See https://learn.microsoft.com/en-us/cpp/c-language/alignment-c?view=msvc-170

> The compiler generally aligns data on natural boundaries that are based on the target processor and the size of the data. Data is aligned on up to 4-byte boundaries on 32-bit processors, and 8-byte boundaries on 64-bit processors.

So a type with a preferred alignment of 8 bytes can be 4-byte aligned on 32-bit targets.

**Comment by saethlin** [comment]
Ah, but `std::mem::align_of` does not return the preferred alignment.

**Comment by ChrisDenton** [comment]
But see also https://learn.microsoft.com/en-us/cpp/build/reference/zp-struct-member-alignment?view=msvc-170

> The compiler stores members after the first one on a boundary that's the smaller of either the size of the member type, or an N-byte boundary.

Where `N` is 8 for x86.

**Comment by glandium** [comment]
It's true that from the perspective that in `struct { int a; uint64_t b; }`, b is 8-bytes aligned, alignof returning 8 is actually correct. But it's more indicative of "you'll be safe if you align to 8 bytes" than "you must align to 8 bytes" or "it will always be aligned to 8 bytes". Because the struct itself, on the stack, will still be 4-bytes aligned. The x86 windows ABI is weird that way... but that essentially breaks the assumption from the alignment check.

**Comment by ChrisDenton** [comment]
Right. But it matters in terms of how this is addressed. It would be wrong to change `align_of` as that's documented to be for struct fields.

For the short term, skipping this check on i686-windows might be the only decent options (which is unfortunate). But longer term it would be better to have a proper fix. I'm not really what that would look like tbh. As you say, it's breaking a core assumption.

**Comment by saethlin** [comment]
> The x86 windows ABI is weird that way... but that essentially breaks the assumption from the alignment check.

There is no assumption made by the alignment check. We unambiguously tell LLVM that a dereference of a `*const u64` on `i686-pc-windows-msvc` is 8-byte-aligned: https://godbolt.org/z/MrdTPqWbY
```rust
pub unsafe extern "C" fn read_u64(ptr: *const u64) -> u64 {
    *ptr
}
```
```llvm
define i64 @_ZN7example8read_u6417hbc6f74c29a983155E(ptr %ptr) unnamed_addr #0 !dbg !5 {
  %0 = load i64, ptr %ptr, align 8, !dbg !10, !noundef !9
  ret i64 %0, !dbg !11
}
```

If you are saying that this alignment assumption is wrong, then we have a soundness issue. The bug is not in the alignment check, we either have invalid codegen from MSVC because it does not sufficiently align its `uint64_t`s, or we have invalid codegen from rustc because it assumes `u64` is 8-byte-aligned on this target when it is not.

**Comment by Mark-Simulacrum** [comment]
cc https://github.com/rust-lang/rust/pull/103830 https://github.com/rust-lang/rust/issues/80127, which I think may also be related in general

**Comment by jyn514** [comment]
Looks like https://github.com/rust-lang/rust/pull/112157 is the latest incarnation of #103830.

**Comment by erikdesjardins** [comment]
#103830 / #112157 is basically the same issue but for byval parameters. The unsoundness is immediately visible because if you assume a byval argument will be 8 byte aligned you'll load from `rsp+8`  instead of `rsp+4`, so Clang (and now rustc) implement a copy of MSVC's alignment logic. But this is not as bad from a language semantics perspective, because byval pointers are not exposed to user code.

Whereas here, for loads of pointer parameters, it looks like we've been generating code that's unsound when linked to MSVC, but it happened to work in most cases.

Here's a comparison of what rustc, Clang, and MSVC generate: https://godbolt.org/z/56h6Md1rd
Clang and rustc both generate the same LLVM IR that assumes 8 byte alignment, and LLVM generates code to align the stack as necessary.

This means we can't reduce our alignment to 4 in all cases, because then we wouldn't have sufficient alignment for FFI calls to Clang-compiled code. Unless we patch Clang at the same time, we'd have to use the higher alignment for our stack/heap allocations, but then use align 4 for loads/stores/etc.

**Comment by deltragon** [comment]
Wouldn't @Gankra's [abi-cafe](https://github.com/Gankra/abi-cafe) project be helpful here?

**Comment by RalfJung** [comment]
Why is it not correct to have `align_of::<u64>` return `4` on these targets? It seems like if C code has a local `uint64_t`, takes a pointer to that, and passes it to Rust, it will in general only be 4-aligned, so any higher value for `align_of` is clearly wrong.

Is it that structs *must* be laid out with the higher alignment? Then we need another internal notion of struct-alignment in the compiler, but the general alignment expected for all references must be 4.

Does this affect only `u64` or any type with alignment > 4?

> This means we can't reduce our alignment to 4 in all cases, because then we wouldn't have sufficient alignment for FFI calls to Clang-compiled code.

To make sure I understand correctly: this is because for values passed on the stack, the alignment is part of the call ABI, so if we use 4 and clang uses 8 we have a problem? Doesn't this already mean that clang is ABI-incompatible with MSVC?

**Comment by ChrisDenton** [comment]
`align_of` is documented like so:

> Returns the [ABI](https://en.wikipedia.org/wiki/Application_binary_interface)-required minimum alignment of a type in bytes.
>
> Every reference to a value of the type T must be a multiple of this number.
>
> This is the alignment used for struct fields. It may be smaller than the preferred alignment.

These two statements are in conflict for 32-bit msvc: "the alignment used for struct fields" and "the ABI-required minimum alignment". There is no way to square that circle, right?

> Does this affect only u64 or any type with alignment > 4?

Any type with an alignment >4 except for when there's an explicit alignment or iirc SIMD types. So in practice it's `u64` or a struct containing a `u64`.

> Doesn't this already mean that clang is ABI-incompatible with MSVC?

Yes.

**Comment by RalfJung** [comment]
> These two statements are in conflict for 32-bit msvc: "the alignment used for struct fields" and "the ABI-required minimum alignment". There is no way to square that circle, right?

I guess it depends on what exactly is meant by "alignment used for struct fields", we can always choose to align struct fields more than that. That would surely be surprising, but it is MSVC that is prescribing this behavior, not us...

> > Doesn't this already mean that clang is ABI-incompatible with MSVC?
>
> Yes.

Sounds like there should also be a clang issue tracking this then?

**Comment by cbeuw** [comment]
> This is the alignment used for struct fields.

rustc cannot possibly be obliged to align struct fields to _precisely_ `align_of`. If an allocation returns an address that happens to have a lot of trailing zeros, the first struct field will have a greater alignment than `align_of`. This happens all the time, and is indistinguishable from "rustc always aligning a struct field to more than `align_of`".

Alignment is a total order, the meaningful question is "Is A aligned to B". It doesn't make sense to talk about equivalence between two alignments, i.e. "does A have _exactly the same_ alignment as B" shouldn't really come up.

It makes sense for `align_of::<u64>` to return 4 on i686 MSVC, and then 

> use the higher alignment for our stack/heap allocations, but then use align 4 for loads/stores/etc.

**Comment by ChrisDenton** [comment]
Rust is obliged to add proper padding to struct fields to align types. If `std::mem::align_of` returns 4 then any code (including in std) that uses `align_of` to calculate padding will now be wrong, at least for `repr(C)`. We would need a new function for padding and to audit the crates.io ecosystem for all the uses of `align_of`.

**Comment by cbeuw** [comment]
> If std::mem::align_of returns 4 then any code (including in std) that uses align_of to calculate padding will now be wrong

This is true, it may be a big change, but it is ultimately a compiler implementation detail.

>audit the crates.io ecosystem for all the uses of `align_of`

While this is certainly prudent, all user code that was previously well-aligned using `align_of` will remain well-aligned. Relaxing the alignment of a type isn't a breaking change.

**Comment by ChrisDenton** [comment]
To be clear it's not just prudent it'd be a necessity. Any code using `align_of` for calaculting layout will now be broken. This is not just an implementation detail of the compiler. People expect to be able to manually calculate the layout of `repr(C)` structs.

We would absolutely need to have a library alternative. Otherwise what are they to replace `align_of` with in that case?

**Comment by cbeuw** [comment]
> People expect to be able to manually calculate the layout of repr(C) structs.

Ah yes, that is a problem.

So in summary, we have 3 options and none are perfect:

1. `align_of` returns 4, actually aligns 8: breaks manual layout calculation for `repr(C)`
2. `align_of` returns 4, actually aligns 4: breaks ABI compatibility with Clang
3. `align_of` returns 8, actually aligns 8 (current behaviour): breaks ABI compatibility with MSVC

**Comment by cbeuw** [comment]
> People expect to be able to manually calculate the layout of repr(C) structs.

Ah yes indeed.

So to summarise, all the options and their consequences are:

|     `align_of` |     Storage alignment |     Store/load alignment |     Consequences                                                          |
|----------------|-----------------------|--------------------------|--------------------------------------------------------------------------|
|     8          |     8                 |     8                    |     (Current behaviour)   breaks ABI-compat with MSVC                    |
|     8          |     8                 |     4                    |     Ok?                                                                  |
| 8              | 4                     | 8                        | Just broken                                                              |
|     8          |     4                 |     4                    |     Breaks ABI-compat with   Clang and `repr(C)` layout   calculation    |
| 4              | 8                     | 8                        | Breaks `repr(C)` layout   calculation                                    |
|     4          |     8                 |     4                    |     Breaks `repr(C)` layout calculation                                  |
|     4          |     4                 |     8                    |     Just broken                                                          |
|     4          |     4                 |     4                    |     Breaks ABI-compat with   Clang                                       |

**Comment by saethlin** [comment]
I do not think the `8 8 4` line is viable. The documentation for `align_of` states

> Every reference to a value of the type `T` must be a multiple of this number.

That means that getting a valid `*const uint64_t` from MSVC code then turning it into a reference could be UB. Fudging our store/load alignment to be different from other meanings of alignment just kicks the UB to somewhere slightly later.

**Comment by cbeuw** [comment]
> The MSVC bug would be the value returned by alignof. That doesn't change anything for the alignment check in rust.

@glandium How did you know this? MS didn't say what the fix is in https://developercommunity.visualstudio.com/t/visual-c-alignof-is-not-conformant-on-x86/1258506. They could just start aligning `uint64_t` to 8 bytes on the stack so everything will be fine.

Also my reading of the comment so far is that `uint64_t` struct fields will continue to be aligned to 8 bytes by MSVC. Surely if MS make `alignof` return 4 then they'd break manual layout calculation in C and C++? So they can't do this.

**Comment by riking** [comment]
"`4 4 4` except it's `4 8 4` inside `#[repr(C)]` structures only" seems like the most correct description of the platform.

If it wasn't for the "you can automatically calculate repr(C) structures using alignof" guarantee, this seems like it would have been a performance-painful but easy fix.

**Comment by thomcc** [comment]
Given how long we've gone without offset_of, I think quite a bit of code is out there that does the automatic calculation as per the definition.

**Comment by nbdd0121** [comment]
Thinking about this again, there is nothing intrinsically unsound with our current behaviour, right? The FFI boundary is already unsafe, so the worse case is that code that takes pointer from C can no longer natively assume that it's aligned, and have to use `addr_of{,_mut}` to safely access the data if the pointee is 8-byte aligned.

To make life a bit easier for accessing these structs, we could also move the alignment requirement (if over 4) on 32-bit MSVC target from validity requirement to safety requirement, and stop compiler and libcore from taking advantage of the alignment. Other safe Rust code can still assume references are aligned to their respective `align_of`, so FFI code needs to take extra care if they let pointers escape to safe code.

**Comment by JakobDegen** [comment]
T-compiler asked for an official response from T-opsem, and there's been enough discussion about this at this point that I think it's worth having an FCP. When this FCP finishes, I'll re-nominate for T-compiler so that the team is aware of this result and can decide on next steps:

The issue: On some targets, MSVC currently fails to sufficiently align values on the stack when their alignment is 8. This is behavior that is incompatible both with assumptions that Rust currently makes and with the C standard.

After discussion, T-opsem does not believe that we should do anything about this on the lang side; #98112 was correct in treating this code as UB. A variety of lang changes to accommodate this behavior were considered ([example](https://github.com/rust-lang/rust/issues/112480#issuecomment-1599055734)), but none of them seem acceptable. We should treat this as a bug in MSVC - there is even a matching report in the [MSVC issue tracker](https://developercommunity.visualstudio.com/t/visual-c-alignof-is-not-conformant-on-x86/1258506) (although that issue has been closed in the meantime).

There has been some discussion about potential MSVC fixes to this bug. It is unclear if there will be any fixes, or what the options under consideration are. If MSVC does anything, opsem/lang *may* reconsider whether a lang change on the Rust side is needed at that time.

T-compiler is free to do damage control to avoid breaking users. This may mean disabling #98112 on the affected targets (as has already been done), reducing the alignment metadata which is emitted to LLVM, documenting the bug, or demoting the affected target to Tier-2. Such changes do not need opsem consultation, but they should also not be seen as language guarantees; they will always be best-effort.

@rfcbot fcp close

(Using "close" since it seems the most appropriate, although the issue should not actually be closed with completion of the FCP)

**Comment by CAD97** [comment]
Just for clarity, I want to specify that this just impacts x86 MSVC; x86_64 MSVC has a 16-byte function call alignment.

Furthermore, I want to point at [Conflicts with the x86 compiler](https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#conflicts-with-the-x86-compiler):

> Data types that are larger than 4 bytes aren't automatically aligned on the stack when you use the x86 compiler to compile an application. Because the architecture for the x86 compiler is a 4 byte aligned stack, anything larger than 4 bytes, for example, a 64-bit integer, can't be automatically aligned to an 8-byte address.
>
> Working with unaligned data has two implications.
>
> - It may take longer to access unaligned locations than it takes to access aligned locations.
> - Unaligned locations can't be used in interlocked operations.
>
> If you require more strict alignment, use __declspec(align(N)) on your variable declarations. This causes the compiler to dynamically align the stack to meet your specifications. However, dynamically adjusting the stack at run time may cause slower execution of your application.

This reinforces the interpretation that this is on the Microsoft Compiler for creating unaligned stack locals, not that they're considered aligned and that Rust is at fault.

Hilariously, [`int64_t` will get a (potentially) 4-byte alignment on the x86 stack](https://c.godbolt.org/z/TTPTdT3GW), but [`_Alignas(int64_t) int64_t` will correctly get an 8-byte alignment](https://c.godbolt.org/z/qMo4sxKce).

**Comment by Trolldemorted** [comment]
>This reinforces the interpretation that this is on the Microsoft Compiler for creating unaligned stack locals, and that it is not the case that they're considered aligned and that Rust is at fault.

Why does this reinforce that interpretation? The explanations given correctly reflect the reality on the ground. x86 will happily read from and write to unaligned locations, and it should be no surprise that interlocked operations have special requirements

>Hilariously, [int64_t will get a (potentially) 4-byte alignment on the x86 stack](https://c.godbolt.org/z/s6G4KvGTb), but [_Alignas(int64_t) int64_t will correctly get an 8-byte alignment](https://c.godbolt.org/z/4GK3WTMv6).

Why is that hilarious? As posted here before this is expected behaviour, an `int64_t` on x86 is [guaranteed to be 4-byte aligned](https://developercommunity.visualstudio.com/t/vs2017-64-bit-int-alignment-problem/294259#T-N298592), but you can force a larger alignment with alignas.

x86 is a very old architecture, and we are interfacing with code that was compiled before some of us were born. These binaries are not getting "fixes" for this alignment, because they won't be rebuilt **and** according to MS this is working as intended.

**Comment by nbdd0121** [comment]
If alignment of int64_t is 4, then `_Alignas(int64_t)` should make something 4-byte aligned, but in reality it forces 8-byte alignment. The point is that MSVC is itself inconsistent.

**Comment by CAD97** [comment]
> Why does this reinforce that interpretation?

Because Microsoft quite literally says "data types [...] aren't automatically aligned on the stack" and calls it "unaligned data." Just as you called them "unaligned locations."

> Why [`_Alignas(int64_t) int64_t`] that hilarious?

Because it's another admission that the x86 compiler is underaligning `int64_t`. Because this is just saying to give me `int64_t` aligned as `int64_t`, but this behaves differently than `int64_t`. It's that asking the compiler to align a type as itself can change the behavior.

The point is that the proper alignment of `int64_t` on x86 is still 8 bytes, not 4, despite them only getting 4 byte alignment on the MSVC compiled stack (unless you ask MSVC to actually align it). You are required to know that this underalignment happens and compensate for it. Everyone agrees that an aligned pointer is 8-byte aligned.

The difference is that x86 MSVC does not conform to the C specification here (which requires objects to only be allocated at an address multiple of the reported `_Alignof`) and avoids this causing issues by refining the standard and allowing pointer reads/writes to unaligned locations (which is UB by the standard, and requires `__unaligned` on x64 MSVC).

Rust references are required to be aligned. If you ask MSVC what that means (i.e. `_Alignas` or `_Alignof`), it will say 8 bytes. So requiring 8 byte alignment is just agreeing with MSVC.

**Comment by glandium** [comment]
The point is we're not trying to win a C standard purity contest here and have a higher standing ground than MSVC. The point is that compatibility with the MSVC ABI has to be respected, and the current behavior of MSVC, however buggy you might think it is, *is* the ABI.

**Comment by ChrisDenton** [comment]
I think the point people are making is that there are challenges involved in supporting a non-conformant ABI.

**Comment by RalfJung** [comment]
If this is the ABI, it is internally inconsistent. A consistent ABIs aligns its variables the same way on the stack and in structs. See above for a discussion of the various ways to fix this -- I don't think a proper fix is possible. (It would require splitting `align_of` into 2 separate pieces of information: struct alignment and "minimal alignment of any valid pointer". Should we require every program on every target to deal with this because one target screwed up?) I would suggest do look into mitigation strategies: basically stop emitting `align` annotations to LLVM on this target. And I guess change the target string we use for LLVM to fix alignment of function arguments passed on the stack? No idea where that is actually controlled.

I wonder what the LLVM folks think about this. After all, clang will have the same problem: a function with many `int64_t` arguments has an ABI mismatch between clang and MSVC.  Could someone please open a clang issue? That requires pointing at the assembly generated by clang and MSVC and saying "here that clearly doesn't match", which is beyond my skills.

**Comment by JakobDegen** [comment]
Nominating for compiler on the basis of the completed FCP [here](https://github.com/rust-lang/rust/issues/112480#issuecomment-1606326864)

**Comment by CAD97** [comment]
> So to summarise, all the options and their consequences are: [[table]](https://github.com/rust-lang/rust/issues/112480#issuecomment-1599055734)

I want to point out since I think it might've been overlooked: [?, 4, ?] (aligning `i64` to 4) doesn't just break ABI with Clang, it breaks ABI with MSVC as well; x86 MSVC aligns `int64_t` to 8 in structures. [[godbolt]](https://c.godbolt.org/z/Waqd3PrPe)

I believe the current x86 MSVC behavior would be [8, 8, 4] or perhaps even [8, 8, 1]. (Relaxing to the latter based purely on vibes of MS documentation treating load alignment as only a performance concern on x86 and [`__unaligned`](https://learn.microsoft.com/en-us/cpp/cpp/unaligned?view=msvc-170) being invalid on x86.)

**Comment by RalfJung** [comment]
According to https://github.com/llvm/llvm-project/issues/34563#issuecomment-1638828061, this bug does not imply an ABI incompatibility... is that something clang somehow achieves? Does rustc need to do anything to get the same ABI?

**Comment by nbdd0121** [comment]
I think that simply means we tell LLVM is i64 is 4-byte aligned but still have the Rust i64 8-byte aligned.

**Comment by Nemo157** [comment]
IIUC https://github.com/rust-lang/rust/pull/112157 fixes the ABI issues, marking `i64` as align-4 when passed by-value.

**Comment by erikdesjardins** [comment]
#112157 fixes the ABI incompatibility where by-value arguments passed on the stack get incorrect stack offsets (meaning we read/write incorrect memory). It does not fix the ABI incompatibility with by-ref/pointer arguments getting incorrect align attributes.

> According to https://github.com/llvm/llvm-project/issues/34563#issuecomment-1638828061, this bug does not imply an ABI incompatibility...

I assume it's been edited since then, but the first word of that comment is "most", so I don't think it's saying that all ABI incompatibilities have been fixed... :P

By "parameter passing isn't affected by the ABI alignment here" they mean the stack offsets of parameters, i.e. the problem fixed by #112157 (which was already implemented in Clang). That problem is easier to fix since byval argument stack slots aren't directly exposed to user code, so they can be underaligned without breaking language semantics.

**Comment by apiraino** [comment]
More discussion in the T-compiler triage meeting ([on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-08-03/near/381510434)), this time floating the idea of a downgrade of the platform in the tier support. Or at least document the peculiarities/quirks of this specific target in our tier support list.

**Comment by eggyal** [comment]
> According to https://github.com/llvm/llvm-project/issues/34563#issuecomment-1638828061, this bug does not imply an ABI incompatibility... is that something clang somehow achieves? Does rustc need to do anything to get the same ABI?

rust-lang/rfcs#3594 proposes to expose an attribute akin to `force_align_arg_pointer` in C compilers, which AIUI clang lowers to the LLVM IR attribute `stackrealign` that in turn injects into the annotated function's preamble a runtime realignment of its stack to 16(?) bytes.

**Comment by jendrikw** [comment]
Related discussion at the GCC bug tracker https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69560

**Comment by RalfJung** [comment]
That's not quite the same. There GCC chose to give types a *higher* alignment to `uint64_t` on the stack than they have in struct fields. They are free to do so as long as they don't indicate that *all* `uint64_t` are aligned -- the bug was that they did so, but it was fixed.

This here is the opposite situation: we have a type that must be 8-aligned, but MSVC decides to make it only 4-aligned when it occurs on the stack. That's just fundamentally broken and there us no sane way to deal with it. We would need a fundamentally new concept, separating a type's alignment for stack layout purposes from its minimal alignment requirement.  We [decided](https://github.com/rust-lang/rust/issues/112480#issuecomment-1606326864) not to do that; all that remains is doing damage control for this fundamentally broken target.

> https://github.com/rust-lang/rfcs/pull/3594 proposes to expose an attribute akin to force_align_arg_pointer in C compilers, which AIUI clang lowers to the LLVM IR attribute stackrealign that in turn injects into the annotated function's preamble a runtime realignment of its stack to 16(?) bytes.

That doesn't help as there will still be code compiled with MSVC that creates pointers that get passed to Rust code.

I think what we should do is
- When generating code for this broken target, never emit `align` annotations to LLVM that are bigger than 4
- Find a suitable place to document this as a "target erratum"

We can even re-enable @saethlin's UB checks for alignment with the same cap of 4, to ensure coherence of the LLVM IR we emit and the UB we check. But Rust crates may still make assumptions based on `align_of` and we're not going to ask the entire ecosystem to deal with the misaligned pointers produced by MSVC.

**Comment by RalfJung** [comment]
@CAD97 @ChrisDenton I realized I am actually unclear about one aspect of this issue: what exactly is the set of types where MSVC reports a higher alignment than what it will use when putting such a variable on the stack? We know `uint64_t` is on that list, but e.g. does it also affect structs containing `uint64_t` types? Does it affect structs explicitly annotation with the C equivalent of `#[repr(align)]`? (I am always assuming the variable declaration itself is not annotated, only the type may have annotations)

**Comment by CAD97** [comment]
> [!WARNING]
> The experimental results referenced below are only my recollection of the behavior I got originally. This description is accurate both to that recollection and the documentation, but I or someone else should triple-check it before trusting it. I hope to do so sometime this week once I have a chance.

@RalfJung: Based on previous experimental results and some fresh crawling of Microsoft Learn, MSVC alignment works basically as so:

- The compiler invocation specifies [`/Zp` (Struct Member Alignment)](https://learn.microsoft.com/en-us/cpp/build/reference/zp-struct-member-alignment?view=msvc-170) to one of `1`, `2`, `4`, `8`, or `16`.
  - If left unspecified, the default is
    - `8` for x86, ARM, and ARM64.
    - `16` for x64 and [ARM64EC](https://learn.microsoft.com/en-us/windows/arm/arm64ec).
  - “Don't use this option unless you have specific alignment requirements.”
  - `#pragma pack` manipulates the same struct member alignment setting with a stateful stack.
- Padding is added between struct members to align them to their size or the current struct member alignment, whichever is *smaller*.
  - For compound types, padding is applied to the greatest alignment of the struct’s members.
- On the stack, data members are aligned to their size or the target's “natural” alignment boundary, whichever is *smaller*, which is `4` for 32-bit targets and `8` for 64-bit targets.
- Global override (yes, *including* overriding any `#pragma pack`): use of `_Alignas`:
  - Declaring a type with `_Alignas(x)` causes any member declaration with that type to be `_Alignas(x)`.
  - Declaring a struct member with `_Alignas(x)` aligns that member. The containing type is implicitly declared with `_Alignas(x)`.
  - Declaring a stack member with `_Alignas(x)` will guarantee that the member is aligned as requested by `x`.
  - Using `_Alignas` to specify a lower alignment than would be provided without the `_Alignas` specifier is ill-formed.
- `_Alignof` reports the alignment used for the type as a struct member. `_Alignas(tag-name)` is equivalent to `_Alignas(_Alignof(tag-name))`.

This indicates that 128-bit primitive data types will have the same alignment issue on x86_64 that 64-bit primitive data types do on x86. Is the abi-cafe checking simd vector type ABI/layout?

**Comment by RalfJung** [comment]
What does this mean for `struct S { uint64_t field; }`? 

> For compound types, padding is applied to the greatest alignment of the struct’s members.

There doesn't seem to be a notion of a type having alignment, only a size and the global alignment flag, so I don't know what this part means.

**Comment by ChrisDenton** [comment]
> does it also affect structs containing uint64_t types?

Yes.

> Does it affect structs explicitly annotation with the C equivalent of `#[repr(align)]`?

No.

You can see this [on godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxBJcpAAOqAqETgwe3r56icmOAkEh4SxRMVxxdpgOqUIETMQE6T5%2BZZj2uQzVtQT5YZHRsbY1dQ2ZZYNdwT1FfaUAlLaoXsTI7BzmAMzByN5YANQma26qLCzBBMTBAHQI%2B9gmGgCC65vbmHsHyKcEAJ7xmAo3d0eD0wqgI0QYOwAbqg8OgdghUMBMBBobCAFQzfZWB6AgD0uJ2oVQBB2YjwwBC6EBqLhpwAbJIIDM9gB2bH3Hacnb0yQAfRJTCxgK58MRyPMdKYmLW7JMLIAIoC8QSiQLaOTKYCFGcvA4dkI1kq2cKuTz%2BTtVEKcQqrfcaTttcRdQReTymaz2SKDaTbSKEUiIBKpba5YqcQ98aT1RTMFSHvazWSKUwFAAOd1yz1cpPMBQQVPMs0C31c/3isyS6Wym3h%2B6RnOxrU6vU5lP55lCMxGrOc04WkM2mXUmFwx3O3mttMZ40PL1mH1D2elsWBivBxePGtAusEhtxu0jh14FjoXksLhmdPMzMmzm8s8X1MLnuigNBqtGsP3DhzWicACsvB%2BNwvCoJwbjWNYDoLEsrzrDwpAEJoP5zAA1iA/4aPonCSEByGkGBHC8AoIBYUhHBaHMcCwEgaAsPEdDROQlB0Qx9AxJCyDxPEvKQlwACcvIGGC2q8qoqZ0nwdBgsQJEQBE%2BERMEtRfJwCFKcwxBfAA8hE2gVORCF0WwgjaQwtCqRRvBYBEXjAG4Yi0CRIGkFgLCGMA4hWa5eDEAZeCQn8%2BEghUXhgmpvCnC0%2BHqhExAqR4WD4Wcx4RaQgXEBESSYPKmDuUY6pGMhcxUAYwAKAAangmAAO7ab8wEIfwggiGI7BSDIgiKCo6jebocQGEVpiWNY%2Bh4BEJGQHMqDxG0JG4tpZgERl5xYJNTK2C0/nOBArjDE0pCBBMhTFFkSQpAI%2B1nTkqTdCd0yba0VRjFdzRPQIHR1HdvQlAMnSvX9X3HT9EhzAoMHLKD2EcIBpDAVoBGcBaEk7CwCicVCAkXMJfwkhAuCECQexmGsXAzLw5GUWhGFYX%2BHC4XD%2BGEcRpGIcVv6cGYeHeczbNWTMcwZckziSEAA). The `and` `sub` pattern is where it's aligning the stack which, as the OP says, is necessary on x86 when alignment is >4.

Also note that SIMD types are special and will be properly aligned.

**Comment by RalfJung** [comment]
What about structs containing SIMD types?

**Comment by ChrisDenton** [comment]
That is aligned. I updated the godbolt link.

**Comment by RalfJung** [comment]
Okay so it's basically entirely behaving as-if u64 and u128 had alignment 4, except that their offset in a struct will be divisible by 8 (also recursively):

Each type has a "required alignment" and a "layout alignment". The former is used to determine which alignment accesses must have (or else UB), the latter is used for struct layout. The required alignment of a struct is the largest required alignment of its fields; the layout alignment of a struct is the largest layout alignment of its fields. Most types have the same value for both kinds of alignment (including the SIMD types), except integer/float primitives larger than 4 have required alignment 4, but layout alignment matching their size.

**Comment by RalfJung** [comment]
> This indicates that 128-bit primitive data types will have the same alignment issue on x86_64 that 64-bit primitive data types do on x86. Is the abi-cafe checking simd vector type ABI/layout?

I tried testing this but `int128_t` is not a valid type...

**Comment by ChrisDenton** [comment]
Right, only SIMD has 128-bit integers for MSVC. There isn't an issue for 64-bit because all "natural" types are <= 8 byte aligned (except for SIMD types which are special). Here's a list of all the built-in types I know of (minus some aliases):

|Type|sizeof|alignof|stack align (32-bit)|
|--|--|--|--|
|`__int8`, `unsigned __int8`, `char`, `bool`|1|1|1|
|`__int16`, `unsigned __int16`, `short`, `unsigned short`, `wchar_t`|2|2|2|
|`__int32`, `unsigned __int32`, `int`, `unsigned int`, `long`, `unsigned long`|4|4|4|
|`float`|4|4|4|
|`double`|8|8|4|
|`__int64`, `unsigned __int64`, `long long`, `unsigned long long`|8|8|4|
|`__ptr32`|4|4|4|
|`__ptr64`|8|8|4|
|`__m64`|8|8|8|
|`__m128`, `__m128d`, `__m128i`|16|16|16|

So all stack alignments for basic types on x86 are capped at 4. There are two exceptions I know of:

1. SIMD types (e.g. `__m64`, `__m128`, etc)
2. The `alignas` attribute will override the cap.

So oddly enough these two structs have different stack alignments but will otherwise be identical:

```cpp
struct A {
    bool b;
    int64_t x;
};
struct B {
    bool b;
    alignas(int64_t) int64_t x;
};
```

**Comment by RalfJung** [comment]
I found this absolute gold in the [MS docs](https://learn.microsoft.com/en-us/cpp/build/x64-software-conventions?view=msvc-170#conflicts-with-the-x86-compiler):

> Data types that are larger than 4 bytes aren't automatically aligned on the stack when you use the x86 compiler to compile an application. Because the architecture for the x86 compiler is a 4 byte aligned stack, anything larger than 4 bytes, for example, a 64-bit integer, can't be automatically aligned to an 8-byte address.

... except that all other compilers got this right, so "can't be automatically aligned" is a plain fabrication. Someone chose to not align or they forgot it.

> Working with unaligned data has two implications. [...]

They are forgetting "interacting with any standards-compliant compiler or with other languages that assume standard C behavior leads to undefined behavior".

**Comment by RalfJung** [comment]
If we end up having a struct layout algorithm that diverts from the current `repr(C)` algorithm for MSVC anyway (to deal with https://github.com/rust-lang/unsafe-code-guidelines/issues/521), I wonder if we could solve this issue by setting the alignment of `u64` to `4` but still ensuring that the offset of a `u64`  field in a struct is always at an 8-aligned offset? In the [table above](https://github.com/rust-lang/rust/issues/112480#issuecomment-1599055734), this was described as "breaks repr(C)", but we have to "break repr(C)" anyway for MSVC targets so we could use that opportunity to properly fix this issue -- maybe?

**Comment by Jules-Bertholet** [comment]
https://learn.microsoft.com/en-us/cpp/c-language/alignment-c says (emphasis mine):

> Use the C11 keyword `_Alignof` to get the ***preferred*** alignment of a type or variable, and `_Alignas` to specify a custom alignment for a variable or user-defined type.
>
> The convenience macros `alignof` and `alignas`, defined in `<stdalign.h>`, map directly to `_Alignof` and `_Alignas`, respectively. These macros match the keywords used in C++. So using the macros instead of the C keywords may be helpful for code portability if you share any code between the two languages.

Let’s compare to the [C standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf#subsubsection.6.5.4.4) (emphasis again mine):

> The `alignof` operator yields the alignment ***requirement*** of its operand type. The operand is not evaluated and the result is an integer constant expression. When applied to an array type, the result is the alignment requirement of the element type.

I think the correct interpretation here is that x86-32 MSVC’s `alignof` is not standard-compliant. MSVC only *requires* `uint64_t` to be aligned to 4 bytes (or maybe even less), but *prefers* it to be aligned to 8. `alignof`/`alignas` on that compiler use this *preferred* alignment, in contravention of the standard, but that does not reflect the minimum alignment *required* to avoid UB.

**Comment by RalfJung** [comment]
It is somewhat ironic that the same section of the docs proclaims some practices that help portability, while also describing a practice that can easily lead to portable code having UB (since portable C code could easily rely on `alignof` returning the *required* alignment of a type for correctness).

**Comment by briansmith** [comment]
> I think the correct interpretation here is that x86-32 MSVC’s `alignof` is not standard-compliant. MSVC only _requires_ `uint64_t` to be aligned to 4 bytes (or maybe even less), but _prefers_ it to be aligned to 8. `alignof`/`alignas` on that compiler use this _preferred_ alignment, in contravention of the standard, but that does not reflect the minimum alignment _required_ to avoid UB.

Note that this is the same for GCC until GCC 8: https://gcc.gnu.org/gcc-8/porting_to.html#alignof

**Comment by mirh** [comment]
FWIW gcc has still its [own bugs](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90569)

**Comment by Jules-Bertholet** [comment]
A "fun" interaction between this issue and other MSVC weirdness: https://github.com/rust-lang/rust/issues/81996#issuecomment-4382558266

**Comment by Jules-Bertholet** [comment]
In fact, thanks to this weirdness, we can more or less polyfill a standard-compliant `alignof` macro in MSVC: https://godbolt.org/z/4M81KGhdM

**Comment by glandium** [comment]
> In fact, thanks to this weirdness, we can more or less polyfill a standard-compliant `alignof` macro in MSVC: https://godbolt.org/z/4M81KGhdM

Does clang-cl output the same thing?

**Comment by Jules-Bertholet** [comment]
> Does clang-cl output the same thing?

Yes.
